flask/app/groups.py

```diff
@@ -0,0 +1,230 @@
+        secret_key=app.config["MINIO_SECRET_KEY"],
+    )
+
+    # Check if a user to be added doesn't exist, 404
+    if "users" in request.json:
+        users = models.User.query.filter(
+            models.User.username.in_(request.json["users"])
+        ).all()
+        # Make sure all users are valid
+        if len(users) != len(request.json["users"]):
+            return "Invalid username provided", 404
+
+        # Clear users from db and minio group
+        if len(group.users) != 0:
+            group_info = minioAdmin.get_group(group_code)
+            for old_user in group_info["members"]:
+                minioAdmin.group_remove(group_code, old_user)
+            group.users[:] = []
+
+        for user in users:
+            # Add to db groups
+            group.users.append(user)
+            # Update/add them to the minio groups
+            minioAdmin.group_add(group_code, user.minio_access_key)
+        # Reset policy for group in case the group did not exist
+        minioAdmin.set_policy(group_code, group=group_code)
+
+    try:
+        db.session.commit()
+        return jsonify(group)
+    except:
+        db.session.rollback()
+        return "Server error", 500
+
+
+@app.route("/api/groups", methods=["POST"])
+@login_required
+@check_admin
+def create_group():
+    if not request.json:
+        return "Request body must be JSON", 400
+
+    group_name = request.json.get("group_name")
+    group_code = request.json.get("group_code")
+
+    if not group_name:
+        return "A group display name must be provided", 400
+    if not group_code:
+        return "A group codename must be provided", 400
+
+    if models.Group.query.filter(
+        (models.Group.group_code == group_code)
+        | (models.Group.group_name == group_name)
+    ).value("group_id"):
+        return "Group already exists", 422
+
+    group_obj = models.Group(group_code=group_code, group_name=group_name)
+
+    # Add users to the db if they were given
+    if request.json.get("users"):
+        users = models.User.query.filter(
+            models.User.username.in_(request.json["users"])
+        ).all()
+        # Make sure all users are valid
+        if len(users) != len(request.json["users"]):
+            return "Invalid username provided", 404
+        for user in users:
+            group_obj.users.append(user)
+
+    minioClient = Minio(
+        app.config["MINIO_ENDPOINT"],
+        access_key=app.config["MINIO_ACCESS_KEY"],
+        secret_key=app.config["MINIO_SECRET_KEY"],
+        secure=False,
+    )
+
+    minioAdmin = MinioAdmin(
+        endpoint=app.config["MINIO_ENDPOINT"],
+        access_key=app.config["MINIO_ACCESS_KEY"],
+        secret_key=app.config["MINIO_SECRET_KEY"],
+    )
+
+    # Create minio bucket via mc, 422 if it already exists
+    if minioClient.bucket_exists(group_code):
+        return "Minio bucket already exists", 422
+
+    minioClient.make_bucket(group_code)
```

Comment:
What happens when the `group_code` contains invalid characters or is too short to be a bucket name? Does the SDK raise an exception?

Reply:
True, both `bucket_exists()` and `make_bucket()` would throw errors which we should definitely be checking for. I've moved those commands closer to the beginning and if they error I return "Invalid bucket name", 422. I've also added a test case to test for this.